Bluetooth: Mesh: Fix lpn groups add when node reset

LingaoM opened the pull request:
Add lpn groups when model sub restore after reset.

Add auto sub `ALL_NODE_ADDR` when provisioned.

Remove all groups when node unprovisoned.

Fixes: #24311
Fixes: #24009

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

jhedberg reviewed (changes requested):
Please move the "all nodes" handling to a separate commit (still same PR though) since these are two distinct issues.

trond-snekvik commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -123,6 +123,10 @@ int bt_mesh_provision(const u8_t net_key[16], u16_t net_idx,
+	if (IS_ENABLED(CONFIG_BT_MESH_LOW_POWER)) {
+		bt_mesh_lpn_group_add(BT_MESH_ADDR_ALL_NODES);
+	}
+
```
IMO, this functionality should be compile time configurable. The working group made it opt-in for a reason, it's very easy to flood the friend queue with broadcast packets that might not be relevant for the LPN

jhedberg replied:
Agreed, and I'd default to disabled for the new option (not just because it seems to be the safe choice but also to keep consistency with previous Zephyr versions).

jhedberg replied:
An alternative would be to expose the group add/del functions as public API (and clearly document that they should only be used for special cases like this, whereas the stack takes automatically care of the rest). Thoughts?

jhedberg replied:
Answering to myself: I'm mildly leaning toward the Kconfig since then the app can't mess around with what the stack manages automatically (e.g. remove groups that the stack had added, etc).

trond-snekvik commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -413,6 +413,14 @@ config BT_MESH_LPN_GROUPS
+
+config BT_MESH_LPN_SUB_ALL_NODE_ADDR
+	bool "Automatically subscribe all node address"
+	default n
```
`default n` is redundant, this line can be removed.

trond-snekvik reviewed (changes requested):
Only one minor comment on the code changes, but please provide a proper commit title and body for your group recovery commit.

jhedberg commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -413,6 +413,14 @@ config BT_MESH_LPN_GROUPS
 	help
 	  Maximum number of groups that the LPN can subscribe to.
+
+config BT_MESH_LPN_SUB_ALL_NODE_ADDR
```
Call it `NODES` instead of `NODE` since that's how it is called in the spec

jhedberg commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -413,6 +413,14 @@ config BT_MESH_LPN_GROUPS
 	  Maximum number of groups that the LPN can subscribe to.
+
+config BT_MESH_LPN_SUB_ALL_NODE_ADDR
+	bool "Automatically subscribe all node address"
```
"nodes" here as well

LingaoM committed:
Bluetooth: Mesh: Add option for lpn auto sub all-nodes

Add option for low power node automatically subscribe
all-nodes-address to friend sub list.

Fixes: #24009

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

LingaoM committed:
Bluetooth: Mesh: Fix add model group address to sub list

When low power node reset, should add group address to
friend subscription list.

Fixes: #24311

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>
